File: samples/sample.java
/**
 * void-space.nvim — Java Sample
 *
 * <p>Demonstrates: classes, interfaces, generics, records, sealed types,
 * enums, lambdas, streams, Optional, annotations, and modern Java patterns.
 */
package space.void.sample;

import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;
import java.util.stream.*;
import java.time.Instant;

// Annotations

@interface Author {
    String name();
    String since() default "1.0";
}

// Enum

enum SpectralClass {
    O(16.0), B(2.1), A(1.4), F(1.04), G(0.8), K(0.45), M(0.0);

    private final double minMass; // solar masses

    SpectralClass(double minMass) {
        this.minMass = minMass;
    }

    static SpectralClass of(double massSolar) {
        SpectralClass[] values = values();
        for (SpectralClass c : values) {
            if (massSolar >= c.minMass) return c;
        }
        return M;
    }

    @Override
    public String toString() { return name(); }
}

// Interface & sealed types

interface CelestialBody {
    String name();
    String describe();

    default String summary() {
        return "[%s] %s".formatted(getClass().getSimpleName(), describe());
    }
}

sealed interface Observation permits VisualObservation, RadioObservation {
    Instant timestamp();
    String target();
    String format();
}

// Records

record VisualObservation(Instant timestamp, String target, double magnitude)
        implements Observation {
    @Override public String format() {
        return "Visual[%s] target=%s mag=%.2f".formatted(timestamp, target, magnitude);
    }
}

record RadioObservation(Instant timestamp, String target, double frequencyGHz)
        implements Observation {
    @Override public String format() {
        return "Radio[%s] target=%s freq=%.3f GHz".formatted(timestamp, target, frequencyGHz);
    }
}

// Custom exception

class ObservationException extends RuntimeException {
    private final String starName;

    ObservationException(String starName, String message) {
        super(message);
        this.starName = starName;
    }

    ObservationException(String starName, String message, Throwable cause) {
        super(message, cause);
        this.starName = starName;
    }

    public String getStarName() { return starName; }
}

// Generic class

class Catalogue<T extends CelestialBody> {
    private final List<T> entries = new ArrayList<>();
    private final Map<String, T> index = new HashMap<>();

    void add(T body) {
        entries.add(body);
        index.put(body.name().toLowerCase(), body);
    }

    Optional<T> find(String name) {
        return Optional.ofNullable(index.get(name.toLowerCase()));
    }

    List<T> all() { return Collections.unmodifiableList(entries); }

    <R> List<R> map(Function<T, R> fn) {
        return entries.stream().map(fn).collect(Collectors.toList());
    }

    List<T> filter(Predicate<T> predicate) {
        return entries.stream().filter(predicate).collect(Collectors.toList());
    }

    @Override
    public String toString() {
        return "Catalogue{size=%d}".formatted(entries.size());
    }
}

// Main class

@Author(name = "void-space", since = "1.0")
public class Sample {

    private static final String VERSION    = "1.0.0";
    private static final double PI         = Math.PI;
    private static final int    MAX_STARS  = 1_000_000;

    private static final Map<String, String> PALETTE = Map.of(
        "void",   "#0d0f18",
        "nebula", "#c792ea",
        "aurora", "#89ddff",
        "comet",  "#82aaff",
        "pulsar", "#ff5370"
    );

    // Inner class

    static class Star implements CelestialBody {
        private final String  name;
        private final double  massSolar;
        private final double  luminosity;
        private final List<String> tags;
        private int observations = 0;

        Star(String name, double massSolar, double luminosity, String... tags) {
            Objects.requireNonNull(name, "name must not be null");
            if (massSolar <= 0) throw new IllegalArgumentException(
                "massSolar must be positive, got " + massSolar);
            this.name        = name;
            this.massSolar   = massSolar;
            this.luminosity  = luminosity;
            this.tags        = List.of(tags);
        }

        @Override public String name() { return name; }

        SpectralClass spectralClass() { return SpectralClass.of(massSolar); }

        void observe() { observations++; }

        int observationCount() { return observations; }

        @Override
        public String describe() {
            return "%s [class %s, mass=%.2f☉, lum=%.2e☉, obs=%d]"
                .formatted(name, spectralClass(), massSolar, luminosity, observations);
        }

        @Override
        public String toString() { return describe(); }
    }

    // Fibonacci (iterative)

    static List<Long> fibonacci(int n) {
        if (n <= 0) return List.of();
        List<Long> seq = new ArrayList<>(n);
        long a = 0, b = 1;
        for (int i = 0; i < n; i++) {
            seq.add(a);
            long tmp = a + b;
            a = b;
            b = tmp;
        }
        return seq;
    }

    // Concurrency

    static List<String> surveyAll(List<Star> stars) throws InterruptedException, ExecutionException {
        ExecutorService exec = Executors.newVirtualThreadPerTaskExecutor();
        try {
            List<Future<String>> futures = stars.stream()
                .map(star -> exec.submit(() -> {
                    Thread.sleep(10);
                    star.observe();
                    return star.describe();
                }))
                .collect(Collectors.toList());

            List<String> results = new ArrayList<>();
            for (Future<String> f : futures) {
                results.add(f.get());
            }
            return results;
        } finally {
            exec.shutdown();
        }
    }

    // Pattern matching helper

    static String classifyObservation(Observation obs) {
        return switch (obs) {
            case VisualObservation v when v.magnitude() < 0 ->
                "Extremely bright visual target: " + v.target();
            case VisualObservation v ->
                "Visual target: " + v.target() + " (mag " + v.magnitude() + ")";
            case RadioObservation r ->
                "Radio source: " + r.target() + " at " + r.frequencyGHz() + " GHz";
        };
    }

    // main

    public static void main(String[] args) throws Exception {
        System.out.printf("void-space v%s — Pi ≈ %.6f%n", VERSION, PI);
        System.out.printf("MAX_STARS = %,d%n%n", MAX_STARS);

        // Stars
        var sirius     = new Star("Sirius",           2.10,  25.4,   "bright", "binary");
        var betelgeuse = new Star("Betelgeuse",       20.0,  1.0e5,  "red-supergiant");
        var proxima    = new Star("Proxima Centauri", 0.12,  0.0017, "nearest");
        var sol        = new Star("Sol",              1.00,  1.0,    "home");

        var stars = List.of(sirius, betelgeuse, proxima, sol);

        // Catalogue + streams
        var catalogue = new Catalogue<Star>();
        stars.forEach(catalogue::add);

        System.out.println("── Catalogue ──");
        catalogue.all().forEach(System.out::println);
        System.out.println();

        // Stream pipeline
        System.out.println("── Massive stars (mass > 5☉) ──");
        catalogue.filter(s -> s.massSolar > 5)
            .stream()
            .sorted(Comparator.comparingDouble((Star s) -> s.massSolar).reversed())
            .map(Star::describe)
            .forEach(System.out::println);
        System.out.println();

        // Names via map
        List<String> names = catalogue.map(Star::name);
        System.out.println("All names: " + String.join(", ", names));

        // Optional
        catalogue.find("sirius")
            .ifPresentOrElse(
                s -> System.out.println("Found: " + s.name()),
                ()  -> System.out.println("Not found"));

        // Statistics
        DoubleSummaryStatistics stats = stars.stream()
            .mapToDouble(s -> s.massSolar)
            .summaryStatistics();
        System.out.printf("%nMass stats — min=%.2f max=%.2f avg=%.2f%n",
            stats.getMin(), stats.getMax(), stats.getAverage());

        // Fibonacci
        List<Long> fibs = fibonacci(10);
        System.out.println("Fibonacci(10): " + fibs);

        // Concurrent survey
        System.out.println("\n── Survey ──");
        List<String> observations = surveyAll(new ArrayList<>(stars));
        observations.forEach(r -> System.out.println(" > " + r));

        // Records & sealed types
        System.out.println("\n── Observations ──");
        List<Observation> obs = List.of(
            new VisualObservation(Instant.now(), "Sirius",  -1.46),
            new RadioObservation(Instant.now(),  "Cygnus A", 1.415)
        );
        obs.forEach(o -> System.out.println(classifyObservation(o)));

        // Palette
        System.out.println("\n── Palette ──");
        PALETTE.entrySet().stream()
            .sorted(Map.Entry.comparingByKey())
            .forEach(e -> System.out.printf("  %-8s %s%n", e.getKey(), e.getValue()));

        // Lambda & method reference demo
        Predicate<Star> isBright = s -> s.luminosity > 100;
        Function<Star, String> formatter = s ->
            "%-20s class=%s".formatted(s.name(), s.spectralClass());

        System.out.println("\n── Bright stars ──");
        stars.stream()
            .filter(isBright)
            .map(formatter)
            .forEach(System.out::println);

        // Exception handling
        try {
            var bad = new Star("", -1, 0);
        } catch (IllegalArgumentException e) {
            System.out.println("\nCaught: " + e.getMessage());
        }

        ObservationException obsEx = new ObservationException("Ghost", "star not found");
        System.out.println("ObservationException for: " + obsEx.getStarName());

        System.out.printf("%nDone. HOME=%s%n", System.getenv("HOME"));
    }
}

File: samples/sample.go
// Package main demonstrates Go syntax for void-space.nvim theme validation.
package main

import (
	"context"
	"errors"
	"fmt"
	"math"
	"os"
	"strings"
	"sync"
	"time"
)

// Constants & vars

const (
	Version    = "1.0.0"
	MaxRetries = 3
	Pi         = math.Pi
)

var (
	ErrNotFound  = errors.New("not found")
	ErrInvalidID = errors.New("invalid id")
)

// Types & interfaces

type SpectralClass string

const (
	ClassO SpectralClass = "O"
	ClassB SpectralClass = "B"
	ClassA SpectralClass = "A"
	ClassF SpectralClass = "F"
	ClassG SpectralClass = "G"
	ClassK SpectralClass = "K"
	ClassM SpectralClass = "M"
)

// CelestialBody is implemented by anything that can be described.
type CelestialBody interface {
	Name() string
	Describe() string
}

// Struct with methods

// Star represents a stellar object.
type Star struct {
	name        string
	massSolar   float64 // in solar masses
	luminosity  float64 // in solar luminosities
	tags        []string
	mu          sync.Mutex
	observations int
}

// NewStar constructs a Star, returning an error on invalid input.
func NewStar(name string, mass, luminosity float64) (*Star, error) {
	if strings.TrimSpace(name) == "" {
		return nil, fmt.Errorf("star name cannot be empty")
	}
	if mass <= 0 {
		return nil, fmt.Errorf("%w: mass=%.2f", ErrInvalidID, mass)
	}
	return &Star{name: name, massSolar: mass, luminosity: luminosity}, nil
}

func (s *Star) Name() string { return s.name }

func (s *Star) SpectralClass() SpectralClass {
	switch {
	case s.massSolar >= 16:
		return ClassO
	case s.massSolar >= 2.1:
		return ClassB
	case s.massSolar >= 1.4:
		return ClassA
	case s.massSolar >= 1.04:
		return ClassF
	case s.massSolar >= 0.8:
		return ClassG
	case s.massSolar >= 0.45:
		return ClassK
	default:
		return ClassM
	}
}

func (s *Star) Describe() string {
	return fmt.Sprintf("%s [class %s, mass=%.2f☉, lum=%.2e☉]",
		s.name, s.SpectralClass(), s.massSolar, s.luminosity)
}

func (s *Star) Observe() {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.observations++
}

// Generics

// Map applies fn to every element of slice and returns the results.
func Map[T, U any](slice []T, fn func(T) U) []U {
	out := make([]U, len(slice))
	for i, v := range slice {
		out[i] = fn(v)
	}
	return out
}

// Filter returns elements for which predicate returns true.
func Filter[T any](slice []T, predicate func(T) bool) []T {
	var out []T
	for _, v := range slice {
		if predicate(v) {
			out = append(out, v)
		}
	}
	return out
}

// Error wrapping

type ObservationError struct {
	StarName string
	Err      error
}

func (e *ObservationError) Error() string {
	return fmt.Sprintf("observation of %q failed: %v", e.StarName, e.Err)
}

func (e *ObservationError) Unwrap() error { return e.Err }

// Goroutines & channels

// surveyStar sends the description of a star into ch, honouring ctx cancellation.
func surveyStar(ctx context.Context, s *Star, ch chan<- string, wg *sync.WaitGroup) {
	defer wg.Done()
	select {
	case <-ctx.Done():
		return
	case <-time.After(10 * time.Millisecond):
		s.Observe()
		ch <- s.Describe()
	}
}

// SurveyAll concurrently surveys all stars and collects results.
func SurveyAll(ctx context.Context, stars []*Star) []string {
	ch := make(chan string, len(stars))
	var wg sync.WaitGroup

	for _, s := range stars {
		wg.Add(1)
		go surveyStar(ctx, s, ch, &wg)
	}

	wg.Wait()
	close(ch)

	var results []string
	for r := range ch {
		results = append(results, r)
	}
	return results
}

// Closures

func makeCounter(start int) func() int {
	n := start
	return func() int {
		n++
		return n
	}
}

// Fibonacci (recursive + memoised)

func fibonacci(n int, memo map[int]int) int {
	if n <= 1 {
		return n
	}
	if v, ok := memo[n]; ok {
		return v
	}
	memo[n] = fibonacci(n-1, memo) + fibonacci(n-2, memo)
	return memo[n]
}

// Palette (map literal)

var palette = map[string]string{
	"void":   "#0d0f18",
	"nebula": "#c792ea",
	"aurora": "#89ddff",
	"comet":  "#82aaff",
	"pulsar": "#ff5370",
}

// main

func main() {
	fmt.Printf("void-space v%s — Pi ≈ %.6f\n", Version, Pi)

	sirius, err := NewStar("Sirius", 2.10, 25.4)
	if err != nil {
		fmt.Fprintf(os.Stderr, "error: %v\n", err)
		os.Exit(1)
	}
	sirius.tags = []string{"bright", "binary", "nearby"}

	betelgeuse, _ := NewStar("Betelgeuse", 20.0, 1e5)
	proxima, _    := NewStar("Proxima Centauri", 0.12, 0.0017)

	stars := []*Star{sirius, betelgeuse, proxima}

	// Print each star
	for _, s := range stars {
		fmt.Println(s.Describe())
	}

	// Generic map/filter
	names   := Map(stars, func(s *Star) string { return s.Name() })
	massive := Filter(stars, func(s *Star) bool { return s.massSolar > 5 })

	fmt.Printf("All names: %v\n", strings.Join(names, ", "))
	fmt.Printf("Massive stars: %d\n", len(massive))

	// Concurrent survey
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	results := SurveyAll(ctx, stars)
	for _, r := range results {
		fmt.Println(" >", r)
	}

	// Closure counter
	next := makeCounter(0)
	for range 5 {
		fmt.Printf("tick %d\n", next())
	}

	// Fibonacci
	memo := make(map[int]int)
	for i := range 10 {
		fmt.Printf("fib(%d)=%d  ", i, fibonacci(i, memo))
	}
	fmt.Println()

	// Palette
	for k, v := range palette {
		fmt.Printf("  %-8s %s\n", k, v)
	}

	// Type assertion & switch
	bodies := []CelestialBody{sirius, betelgeuse}
	for _, b := range bodies {
		switch v := b.(type) {
		case *Star:
			fmt.Printf("Star %q has %d observations\n", v.Name(), v.observations)
		default:
			fmt.Printf("Unknown body: %T\n", v)
		}
	}

	// Multi-return & error check
	_, badErr := NewStar("", -1, 0)
	var obsErr *ObservationError
	wrappedErr := &ObservationError{StarName: "Ghost", Err: ErrNotFound}
	fmt.Println("bad star:", badErr)
	fmt.Println("wrapped:", wrappedErr)
	fmt.Println("is NotFound?", errors.Is(wrappedErr, ErrNotFound))
	fmt.Println("as ObsErr?", errors.As(wrappedErr, &obsErr))

	// Deferred + slice ops
	numbers := []int{3, 1, 4, 1, 5, 9, 2, 6, 5, 3}
	sum := 0
	for _, n := range numbers {
		sum += n
	}
	fmt.Printf("sum(%v) = %d\n", numbers, sum)

	fmt.Printf("HOME=%s\n", os.Getenv("HOME"))
}
